renderlib/MathUtil.h

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include "glm.h"
+
+struct LinearSpace3f
+{
+  glm::vec3 vx;
+  glm::vec3 vy;
+  glm::vec3 vz;
+
+  LinearSpace3f()
+    : vx(1.0f, 0.0f, 0.0f)
+    , vy(0.0f, 1.0f, 0.0f)
+    , vz(0.0f, 0.0f, 1.0f)
+  {
+  }
+  LinearSpace3f(glm::vec3 vx, glm::vec3 vy, glm::vec3 vz)
+    : vx(vx)
+    , vy(vy)
+    , vz(vz)
+  {
+  }
+
+  LinearSpace3f inverse() const;
+
+private:
+  float determinant() const;
+  LinearSpace3f transpose() const;
+  LinearSpace3f adjoint() const;
+};
+
+struct AffineSpace3f
+{
+public:
+  struct LinearSpace3f l;
+  glm::vec3 p;
+
+  AffineSpace3f()
+    : l()
+    , p()
+  {
+  }
+  AffineSpace3f(struct LinearSpace3f l, glm::vec3 p)
+    : l(l)
+    , p(p)
+  {
+  }
+  AffineSpace3f inverse() const;
+};
+
+inline glm::vec3
+xfmVector(const struct LinearSpace3f& xfm, glm::vec3 p)
+{
+  return xfm.vx * p.x + xfm.vy * p.y + xfm.vz * p.z;
+}
+inline glm::vec3
+xfmVector(const struct AffineSpace3f& xfm, glm::vec3 p)
+{
+  return xfmVector(xfm.l, p) + xfm.p;
+}
+
+/// @brief Intersect a line with a plane and return the parametric distance along the
+/// line. The plane is expressed in vector form such as for any point P on the plane
+/// this condition is true:
+///     dot((P - planeP), planeN) = 0
+/// The line is epressed in parametric form:
+///     P = lineP + lineV * lineT
+///
+/// This function solves for distance lineT.
+inline float
+linePlaneIsectT(const glm::vec3& planeP, const glm::vec3& planeN, const glm::vec3& lineP, const glm::vec3& lineV)
+{
+  float det = dot(lineV, planeN);
+  float lineT = (abs(det) < 1e-19f ? 0.0f : dot(planeP - lineP, planeN) / det);
+
+  return lineT;
+}
```

Comment:
Is lineV expected to be normalized?

Reply:
It doesn't need to be normalized because the returned value is a "t" value that is intended to be multiplied with lineV.  I think in this case the planeN is expected to be normalized.